pkg/driver/controller.go

```diff
@@ -78,6 +105,167 @@ func getRoundedCapacity(size int64) int64 {
+				"failed to delete volume %v: %v", vol.GetName(), err)
+		}
+		return vol, true, status.Error(codes.ResourceExhausted, errMsg)
+	}
+
+	return vol, false, status.Error(codes.Aborted, errMsg)
+}
+
+func (cs *controller) init() error {
+	cfg, err := k8sapi.Config().Get()
+	if err != nil {
+		return errors.Wrapf(err, "failed to build kubeconfig")
+	}
+
+	kubeClient, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return errors.Wrap(err, "failed to build k8s clientset")
+	}
+
+	openebsClient, err := clientset.NewForConfig(cfg)
+	if err != nil {
+		return errors.Wrap(err, "failed to build openebs clientset")
+	}
+
+	kubeInformerFactory := kubeinformers.NewSharedInformerFactory(kubeClient, 0)
+	openebsInformerfactory := informers.NewSharedInformerFactoryWithOptions(openebsClient,
+		0, informers.WithNamespace(device.DeviceNamespace))
+
+	// set up signals so we handle the first shutdown signal gracefully
+	stopCh := signals.SetupSignalHandler()
+
+	cs.k8sNodeInformer = kubeInformerFactory.Core().V1().Nodes().Informer()
+	cs.deviceNodeInformer = openebsInformerfactory.Local().V1alpha1().DeviceNodes().Informer()
+
+	if err = cs.deviceNodeInformer.AddIndexers(map[string]cache.IndexFunc{
+		LabelIndexName(cs.indexedLabel): LabelIndexFunc(cs.indexedLabel),
+	}); err != nil {
+		return errors.Wrapf(err, "failed to add index on label %v", cs.indexedLabel)
+	}
+
+	go cs.k8sNodeInformer.Run(stopCh)
+	go cs.deviceNodeInformer.Run(stopCh)
+
+	// wait for all the caches to be populated.
+	klog.Info("waiting for k8s & device node informer caches to be synced")
+	cache.WaitForCacheSync(stopCh,
+		cs.k8sNodeInformer.HasSynced,
+		cs.deviceNodeInformer.HasSynced)
+	klog.Info("synced k8s & device node informer caches")
+	return nil
+}
+
+// CreateDeviceVolume create new device volume for csi volume request
+func CreateDeviceVolume(ctx context.Context, req *csi.CreateVolumeRequest,
+	params *VolumeParams) (*apis.DeviceVolume, error) {
+	volName := strings.ToLower(req.GetName())
+	capacity := strconv.FormatInt(getRoundedCapacity(
+		req.GetCapacityRange().RequiredBytes), 10)
+
+	vol, err := device.GetDeviceVolume(volName)
+	if err != nil {
+		if !k8serror.IsNotFound(err) {
+			return nil, status.Errorf(codes.Aborted,
+				"failed get device volume %v: %v", volName, err.Error())
+		}
+		vol, err = nil, nil
+	}
+
+	if vol != nil {
+		if vol.DeletionTimestamp != nil {
+			if err = device.WaitForDeviceVolumeDestroy(ctx, volName); err != nil {
+				return nil, err
+			}
+		} else {
+			if vol.Spec.Capacity != capacity {
+				return nil, status.Errorf(codes.AlreadyExists,
+					"volume %s already present", volName)
+			}
+			var reschedule bool
+			vol, reschedule, err = waitForDeviceVolume(ctx, vol)
+			// If the device volume becomes ready or we can't reschedule failed volume,
+			// return the err.
+			if err == nil || !reschedule {
+				return vol, err
+			}
+		}
+	}
+
+	nmap, err := getNodeMap(params.Scheduler, params.DeviceName)
```

Comment:
this is not needed for device localpv.

Reply:
as we are going to support waitforfirstconsumer only, but we can keep this is plan is to support immediate mode.